Automattic/wp-calypso
# Reader tracking events - fix a couple bugs related to the general ui_algo prop.

Related to #

## Proposed Changes

* Fixes an issue where `calypso_reader_article_opened` did not supply the correct value of `ui_algo` for various versions of the discover feed, and would always populate this as coming from the `recommended` tab from discover.
* Fixes an issue where various stream `*_loaded` and `*_viewed` events were inconsistent with their `ui_algo` value, due to relying on the path being set on the window (which may not always happen before the event fires).

More details about the two bugs and their fixes below:

### Bug 1 - `calypso_reader_article_opened`

Not a regression, but a short-sight recent additions did not cover.

This `calypso_reader_article_opened` event was not showing the proper `ui_algo` prop when it comes to the discover stream. This event uses the pre-existing `pathnameOverride` to ensure the `ui_algo` is populated WRT the stream the user opened this article from. However, our [recent addition](https://github.com/Automattic/wp-calypso/pull/82144) did not account for this and looked at the current window location instead of the override for providing more specific discover info. This resulted all discover streams populating the prop as `discover_recommended` for this event, even if they were from other discover tabs.

We fix this by ensuring the pathnameOverride is used for finding this extra info when present. Now the `getLocation` function will check the path that is passed to it instead of the window, and the window's version continues to be the fallback passed to the function in the case that the override is not present.

To repro the original bug:
* go to the reader
* navigate to a tab in discover that is not 'recommended'
* click a post card to open the full post in the reader
* notice the event `calypso_reader_article_opened` has a value of `discover_recommended` for `ui_algo` even though we did not come from the recommended tab.

### Bug 2 - various `*_loaded`/`*_viewed` events showing inconsistent values

This bug seems to have existed for quite a long time.

When loading a stream, there is often a `*_viewed` or `*_loaded` event. I found there are inconsistent values for `ui_algo` when loading a stream. This seems to be a race condition present since new location on the window may not be set before the event fires. 

To mitigate this inconsistency, I have added a `pathnameOverride` prop to these specific events that is populated with the value from `getCurrentRoute` on calypso state, since this is reliably set by the time the event fires.

To repro the original bug (taking Discover as an example):
* navigate to the reader landing on 'recent'
* click the discover menu item
* notice tracking event for viewing this has `ui_algo: discover_recommended`
* <img width="490" alt="Screenshot 2023-10-03 at 10 31 47 AM" src="https://github.com/Automattic/wp-calypso/assets/28742426/90476cb7-501e-43c1-aa3b-31c58ea6bb9e">
* click back to 'recent'
* click back to discover again.
* notice the tracking event for viewing this now has `ui_algo: following`.
* <img width="543" alt="Screenshot 2023-10-03 at 10 32 04 AM" src="https://github.com/Automattic/wp-calypso/assets/28742426/c86fe724-f875-4d3a-96ba-94200b9ef442">

## Testing Instructions



* Load this calypso build.
* Try to repro bug 1 with steps noted above. Verify you now see a value corresponding to the discover tab you came from.
* Try to repro bug 2 with steps noted above. Verify the `ui_algo` prop's value is now consistent and corresponds to the stream you loaded. Test this for various feeds.
* Test that there are no regressions with `calypso_reader_site_followed`/`*_unfollowed` (as they also use the getLocation function we udpated) by clicking the follow blog button in the post card ellipsis menu.

## Pre-merge Checklist
